### ISSUE TITLE
Session start

### ISSUE THREAD

**Issue by andrii-pukhalevych** [cakephp/cakephp#14707]
This is a (multiple allowed):

* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

Can I do PR in master to use`$this->started()` instead of `$this->_started` in [`Cake\Http\Session::start()`](https://github.com/cakephp/cakephp/blob/2ebb00a9f87261cd6ba5930e9ca005c882bbb9a3/src/Http/Session.php#L336) ?
Because now it throws `RuntimeException('Session was already started')` if session was manually started (it happens because Im using cakephp integration with some cms).

**Comment by andrii-pukhalevych** [comment]
I want to have this in 3.9

**Comment by andrii-pukhalevych** [comment]
Maybe also remove `'cookie' => 'CAKEPHP'` for php sessions in 3.9?

**Comment by markstory** [comment]
> Maybe also remove 'cookie' => 'CAKEPHP' for php sessions in 3.9?

Changing the default will break compatibility and existing applications hard. If you need a different cookie name you should be able to configure one in your session configuration.

> Can I do PR in master to use$this->started() instead of $this->_started in Cake\Http\Session::start() ?

I think you would need to replace all the `$this->_started` conditions to have a proper change, as `close()` would behave incorrectly if only `start()` was changed.

While changing the check would avoid the exception, it would also allow the session to be configured in a way that is incompatible with the configuration provided to CakePHP. For example your CMS application could be configured to use file sessions, whilst Cake is configured to store sessions in redis. When sessions are not stored in redis it looks like CakePHP is broken. Beyond storage problems, configuration options related to session security would not be applied and could be 'wrong'.

**Comment by andrii-pukhalevych** [comment]
PR is ready, tests were passed

**Comment by ADmad** [comment]
Seems you simply ignored the last paragraph of @markstory's reply.

**Comment by ADmad** [comment]
@markstory Could we add a `ServerRequest::setRequest()/withRequest()` or an additional argument to `ServerRequestFactory::fromGlobals()` for passing session instance, so that users can easily create request instances with instances of customized `Session` class?

**Comment by andrii-pukhalevych** [comment]
Actually I not fully follow last paragraph. Does it mean that `$this->started()` can not be used in `start()`?

**Comment by andrii-pukhalevych** [comment]
Maybe create custom session engine that will do the stuff or add some option to set `_started` to true for php sessions if `session_status() === \PHP_SESSION_ACTIVE` and disable session close to allow external cms handle start and close?
`_started` has protected access and there is no way to ommit its check.

**Comment by markstory** [comment]
> Actually I not fully follow last paragraph. Does it mean that $this->started() can not be used in start()?

The point I was trying to make was that an externally started session and the session configuration defined in a cakephp application could diverge, and there is no way for cakephp to update the configuration (because the session has been started) and raising an exception is effectively what the code does today.

Perhaps this is an acceptable caveat/problem for folks that are using externally configured sessions.

> ServerRequestFactory::fromGlobals() for passing session instance, so that users can easily create request instances with instances of customized Session class?

Yes this could work and makes all the tradeoffs more explicit to those that need it. Another related approach would be to allow the session object to be replaced with  an implementation that doesn't check started state or attempt to configure the session.

**Comment by andrii-pukhalevych** [comment]
I extended session 
```
class Session extends \Cake\Http\Session {

	public function __construct(array $config = []) {
		parent::__construct($config);
		$this->_started = headers_sent() || session_status() === PHP_SESSION_ACTIVE;
	}
}
```

And inserted it into request using deprecated method in middleware, because there is now way to pass session into `ServerRequestFactory::fromGlobals()`:
```
public function middleware($middleware) {
	$middleware->add(function (ServerRequest $request, ResponseInterface $response, $next) {
		$Session = new \App\Utility\Session();
		@$request->session($Session);
		return $next($request, $response);
	});
```

**Comment by andrii-pukhalevych** [comment]
Closed #14710, so please wrap it in correct way.

**Comment by markstory** [comment]
I took a look into adding a new method to replace the session and remembered/re-discovered that the session can be replaced using `$request->withAttribute('session', ...)` This behavior exists in both 3.x and 4.x. Given that we have an existing method I don't see a need to add another way.
